File: 2015/AoC15-16ab.swift
struct Aunt {
  let num: Int
  let data: [(String, Int)]

  init(_ num: Int, _ data: [(String, Int)]) {
    self.num = num
    self.data = data
  }

  static func process(_ line: String) -> Aunt {
    let temp = line.filter( { $0 != " " } ).split(separator: ":", maxSplits: 1)

    let num: Int = Int(temp[0].filter( { $0.isNumber } ))!

    let data: [(String, Int)] = temp[1].split(separator: ",")
      .map( {
        let s = $0.split(separator: ":")
        return (String(s[0]), Int(s[1])!)
      } )

    return Aunt(num, data)
  }
}

let targetData: [String: (Int, (Int) -> Bool)] = [
  "children":    (3, { $0 == 3 } ),
  "cats":        (7, { $0  > 7 } ),
  "samoyeds":    (2, { $0 == 2 } ),
  "pomeranians": (3, { $0  < 3 } ),
  "akitas":      (0, { $0 == 0 } ),
  "vizslas":     (0, { $0 == 0 } ),
  "goldfish":    (5, { $0  < 5 } ),
  "trees":       (3, { $0  > 3 } ),
  "cars":        (2, { $0 == 2 } ),
  "perfumes":    (1, { $0 == 1 } )
]

var aunts = [Aunt]()
while let line = readLine() {
  aunts.append(Aunt.process(line))
}

let ans1 = aunts.first(where: {
  $0.data.allSatisfy( {
    let (key, val) = $0
    return targetData[key]!.0 == val
  } )
} )!.num

let ans2 = aunts.first(where: {
  $0.data.allSatisfy( {
    let (key, val) = $0
    return targetData[key]!.1(val)
  } )
} )!.num

print("Part 1 answer: \(ans1)")
print("Part 2 answer: \(ans2)")

File: 2015/AoC15-16ab.rs
use std::fs;
use std::collections::HashMap;

struct Aunt {
  num: u32,
  data: Vec<(String, u32)>
}

impl From<&str> for Aunt {
  fn from(line: &str) -> Aunt {
    let temp = line.split_whitespace().collect::<String>();
    let (a, b) = temp.split_once(':').unwrap();

    let num: u32 = a.chars()
    .filter(|c| c.is_digit(10))
    .collect::<String>()
    .parse().unwrap();

    let data: Vec<(String, u32)> = b.split(',')
    .map(|s| {
      let (key, value) = s.split_once(':').unwrap();
      (String::from(key), value.parse().unwrap())
    }).collect();

    Aunt { num, data }
  }
}

fn main() {
  let target_data = HashMap::from([
    ("children",    (3, (|x| x == 3) as fn(u32) -> bool)),
    ("cats",        (7, (|x| x  > 7))),
    ("samoyeds",    (2, (|x| x == 2))),
    ("pomeranians", (3, (|x| x  < 3))),
    ("akitas",      (0, (|x| x == 0))),
    ("vizslas",     (0, (|x| x == 0))),
    ("goldfish",    (5, (|x| x  < 5))),
    ("trees",       (3, (|x| x  > 3))),
    ("cars",        (2, (|x| x == 2))),
    ("perfumes",    (1, (|x| x == 1)))
  ]);


  let input = fs::read_to_string("input.txt").unwrap();
  let aunts: Vec<Aunt> = input.split('\n').map(|line| Aunt::from(line)).collect();

  let ans1 = aunts.iter().find(|aunt|
    aunt.data.iter().all(|(key, value)| target_data[key.as_str()].0 == *value)
  ).unwrap().num;
  println!("Part 1 answer: {ans1}");

  let ans2 = aunts.iter().find(|aunt|
    aunt.data.iter().all(|(key, value)| target_data[key.as_str()].1(*value))
  ).unwrap().num;
  println!("Part 2 answer: {ans2}");
}